compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1129,6 +1145,115 @@ public void visit(BLangTypeInit cIExpr) {
+                // member is not an object.
+                continue;
+            }
+            if ((memberType.tsymbol.flags & Flags.ABSTRACT) == Flags.ABSTRACT) {
+                dlog.error(cIExpr.pos, DiagnosticCode.CANNOT_INITIALIZE_ABSTRACT_OBJECT, lhsUnionType.tsymbol);
+                cIExpr.initInvocation.argExprs.forEach(expr -> checkExpr(expr, env, symTable.noType));
+            }
+
+            BAttachedFunction initializerFunc = ((BObjectTypeSymbol) memberType.tsymbol).initializerFunc;
+            if (isArgsMatchesFunction(cIExpr.argsExpr, initializerFunc)) {
+                matchingLhsMemberTypes.add(memberType);
+            }
+        }
+        return matchingLhsMemberTypes;
+    }
+
+    private BType getMatchedTypeOrEmitError(List<BType> matchingLhsMembers, BLangTypeInit cIExpr, BType lhsUnion) {
+        if (matchingLhsMembers.isEmpty()) {
+            // No union type member found which matches with initializer expression.
+            dlog.error(cIExpr.pos, DiagnosticCode.CANNOT_INFER_OBJECT_TYPE_FROM_LHS, lhsUnion);
+            resultType = symTable.semanticError;
+            return symTable.semanticError;
+        } else if (matchingLhsMembers.size() == 1) {
+            // We have a correct match.
+            return matchingLhsMembers.get(0).tsymbol.type;
+        } else {
+            // Multiple matches found.
+            dlog.error(cIExpr.pos, DiagnosticCode.AMBIGUOUS_TYPES, lhsUnion);
+            resultType = symTable.semanticError;
+            return symTable.semanticError;
+        }
+    }
+
+    private boolean isArgsMatchesFunction(List<BLangExpression> invocationArguments, BAttachedFunction function) {
+        if (function == null) {
+            return invocationArguments.isEmpty();
+        }
+
+        if (function.symbol.params.isEmpty() && invocationArguments.isEmpty()) {
+            return true;
+        }
+
+        List<BLangExpression> namedArgs = new ArrayList<>();
+        List<BLangExpression> unnamedArgs = new ArrayList<>();
+        for (BLangExpression argument : invocationArguments) {
+            if (argument.getKind() == NodeKind.NAMED_ARGS_EXPR) {
+                namedArgs.add(argument);
+            } else {
+                unnamedArgs.add(argument);
+            }
+        }
+
+        // All named arguments must be present in function as defaultable parameters.
+        if (!matchDefaultableParameters(function, namedArgs)) {
+            return false;
+        }
+
+        // Given arguments are less than required parameters.
+        int requiredParamCount = function.symbol.params.size();
+        if (requiredParamCount > unnamedArgs.size()) {
+            return false;
+        }
+
+        // No rest params, all (unnamed) args must match params.
+        if (function.symbol.restParam == null && requiredParamCount != unnamedArgs.size()) {
+            return false;
+        }
+
+        // Each arguments must be assignable to required parameter at respective position.
```

Comment:
order doesn't matter for named-args. if the user pass the named-args in a different order than the function definition, will it get handled here?

Reply:
It's handled in `matchDefaultableParameters`.
We can't pass arguments to ordinary parameters by name right?

Reply:
Ack. 
btw, where do we check the restParam Type?

Reply:
Yes after choosing a member type from union we run `types.checkType(cIExpr, matchedType, actualType);`

Reply:
shouldn't we check the rest param when choosing a member? 
Lets say `Foo` has a constructor `__init(string... x)` and `Bar` has a constructor `__init(int... x)`
then `Foo|Bar y = new(5)`. What would be the output?

Reply:
Yes we need to consider that, I completely missed that scenario.